_source/_posts/2022-07-08-spa-web-security-csrf-xss.md

```diff
@@ -0,0 +1,113 @@
+
+|Posts in the SPA web security series|
+| --- |
+| 1. [Defend Your SPA from Security Woes](/blog/2022/07/06/spa-web-security) |
+| 2. [Defend Your SPA from Common Web Attacks](/blog/2022/07/08/spa-web-securty-csrf-xss) |
+
+ Ready to continue and make your SPA safer? This post will use the concepts we introduced to banish some well-known web vulnerabilities.
+
+ {% include toc.md %}
+
+## Practice data cleanliness
+Cross-Site Scripting (XSS) is a vulnterability that continues to plague web developers. This vulnerability is a type of injection attack and is so common that it's [number 3 on the OWASP Top Ten list](https://owasp.org/Top10/A03_2021-Injection/). 
+
+How does this vulnerability work? In short, XSS happens when code pollutes data, and you don't implement safeguards. 
+
+Ok, fine, but what does that mean? A classic example is a website that allows user input, like adding comments or reviews. Suppose the input takes anything the user types in and doesn't appropriately safeguard that user input before storing the input and displaying it to other users. In that case, it's at risk for XSS. 
+
+Imagine an overly dramatic but otherwise innocent scenario like this
+>1. A website allows you to add comments about your favorite K-Drama.
+>2. An agitator adds the comment `<script>alert('Crash Landing on You stinks!');</script>`.
+>3. That terrible comment saves as is to the database.
+>4. A K-Drama fan opens the website.
+>5. The terrible comment is added to the website, appending the `<script></script>` tag to the DOM.
+>6. The K-Drama fan is outraged by the JavaScript alert saying their favorite K-Drama stinks.
+
+So, this is obviously awful. Least of all because we all know "[Crash Landing on You](https://www.imdb.com/title/tt10850932/)" is, in fact, wonderful, but also, this scenario exposes a colossal breach in what injection attacks can do. With JavaScript, the attack could grab cookies, dig around your browser's local storage for authentication information, access your file system, make calls to other sites, and do even more harm.
+
+In this case, we show an example very clearly using a `<script></script>` tag just to make it easier for us to talk about, but the attack can use anything that runs JavaScript, such as adding an HTML element with JS embedded in the attribute, adding JS to resource URLs, embedding JS into CSS, and so forth. This is the true stuff of nightmares!
+
+How do we avoid this sort of shenanigans? There are a few different things to do, but the main one is to ensure you practice good data hygiene through **escaping** and **sanitizing**.
+
+Escaping (where you replace certain characters in HTML so it's as text, such as replacing `<b>` with `&lt;b&gt;`). When you escape, the browser no longer treats the value as part of the code. It's simply text or data. While we're talking XSS here, escaping is a good practice in general as it will protect you against SQL injection.
```

Comment:
```suggestion
*Escaping* refers to replacing certain specific characters in HTML so they are read as text strings, such as replacing `<b>` with `&lt;b&gt;`). When you escape, the browser no longer treats the value as part of the code. It's simply text or data. Although we're talking XSS here, escaping is a good practice in general as it will protect you against SQL injection, a different type of injection attack.
```
please make sure my edits are still accurate